docs/src/components/ComponentDoc/ComponentSourceManager/ComponentSourceManager.ts

```diff
@@ -0,0 +1,138 @@
+import * as _ from 'lodash'
+import * as React from 'react'
+
+import { ExampleSource } from 'docs/src/types'
+import formatCode from 'docs/src/utils/formatCode'
+import { componentAPIs as APIdefinitions, ComponentAPIs } from './componentAPIs'
+import getExampleSource from './getExampeSource'
+
+export type ComponentSourceManagerRenderProps = ComponentSourceManagerState & {
+  handleCodeAPIChange: (newApi: keyof ComponentAPIs) => void
+  handleCodeChange: (newCode: string) => void
+  handleCodeFormat: () => void
+  handleCodeLanguageChange: (newLanguage: ComponentSourceManagerLanguage) => void
+  handleCodeReset: () => void
+}
+
+export type ComponentSourceManagerLanguage = 'js' | 'ts'
+
+export type ComponentSourceManagerProps = {
+  examplePath: string
+  children: (renderProps: ComponentSourceManagerRenderProps) => React.ReactNode
+}
+
+type ComponentSourceManagerAPIs = ComponentAPIs<{
+  sourceCode: ExampleSource | undefined
+  supported: boolean
+}>
+
+export type ComponentSourceManagerState = {
+  currentCodeLanguage: ComponentSourceManagerLanguage
+  currentCodeAPI: keyof ComponentAPIs
+  currentCodePath: string
+
+  componentAPIs: ComponentSourceManagerAPIs
+  currentCode?: string
+  formattedCode?: string
+  originalCode?: string
+
+  canCodeBeFormatted: boolean
+  wasCodeChanged: boolean
+}
+
+export default class ComponentSourceManager extends React.Component<
+  ComponentSourceManagerProps,
+  ComponentSourceManagerState
+> {
+  constructor(props: ComponentSourceManagerProps) {
+    super(props)
+
+    const componentAPIs = _.mapValues(APIdefinitions, (definition, name: keyof ComponentAPIs) => {
+      const sourceCode = getExampleSource(props.examplePath, name)
+
+      return {
+        ...definition,
+        sourceCode,
+        supported: !!sourceCode,
+      }
+    }) as ComponentSourceManagerAPIs
+
+    this.state = {
+      currentCodeLanguage: 'js' as ComponentSourceManagerLanguage,
+      currentCodeAPI: _.findKey(componentAPIs, { supported: true }) as keyof ComponentAPIs,
+      currentCodePath: '',
+
+      componentAPIs,
+      canCodeBeFormatted: false,
+      wasCodeChanged: false,
+    }
+  }
+
+  static getDerivedStateFromProps(
+    props: ComponentSourceManagerProps,
+    state: ComponentSourceManagerState,
+  ): Partial<ComponentSourceManagerState> {
+    const { examplePath } = props
+    const { componentAPIs, currentCodeAPI, currentCodeLanguage, currentCode: storedCode } = state
+
+    const sourceCodes = componentAPIs[currentCodeAPI].sourceCode
+    const originalCode = sourceCodes[currentCodeLanguage]
+
+    const currentCode = storedCode || originalCode
+    const currentCodePath = examplePath + componentAPIs[currentCodeAPI].fileSuffix
+
+    const codeParser = currentCodeLanguage === 'ts' ? 'typescript' : 'babylon'
```

Comment:
It seems that this doesn't raise any error in the Typescript mode:
```
const notString: string = 3
```

Reply:
Yep, we don't have TS checks there. And I'm not sure that we actually need them for now.

I renamed `codeParser` to `prettierParser` to avoid confusion

Reply:
Got it now, I thought it was actually a code parser. I know that the Monaco editor was configurable to parse and show ts errors, so I thought there is something similar here.